File: app/src/main/java/me/zhanghai/android/files/provider/common/LinkOptions.java
/*
 * Copyright (c) 2018 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.provider.common;

import java.util.Objects;

import androidx.annotation.NonNull;
import java8.nio.file.LinkOption;

public class LinkOptions {

    private LinkOptions() {}

    public static boolean hasNoFollowLinks(@NonNull LinkOption... options) {
        boolean noFollowLinks = false;
        for (LinkOption option : options) {
            Objects.requireNonNull(option);
            if (option == LinkOption.NOFOLLOW_LINKS) {
                noFollowLinks = true;
            } else {
                throw new UnsupportedOperationException(option.toString());
            }
        }
        return noFollowLinks;
    }
}

File: app/src/main/java/me/zhanghai/android/files/provider/common/LinkOptions.kt
/*
 * Copyright (c) 2018 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.provider.common

import java8.nio.file.LinkOption

class LinkOptions(val noFollowLinks: Boolean) {
    fun toArray(): Array<LinkOption> {
        val options = mutableListOf<LinkOption>()
        if (noFollowLinks) {
            options += LinkOption.NOFOLLOW_LINKS
        }
        return options.toTypedArray()
    }
}

fun Array<out LinkOption>.toLinkOptions(): LinkOptions {
    var noFollowLinks = false
    for (option in this) {
        when (option) {
            LinkOption.NOFOLLOW_LINKS -> noFollowLinks = true
            else -> throw UnsupportedOperationException(option.toString())
        }
    }
    return LinkOptions(noFollowLinks)
}
